src/Microsoft.VisualStudio.Editor.Razor/DefaultEditorSettingsManagerInternal.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Editor;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    internal class DefaultEditorSettingsManagerInternal : EditorSettingsManagerInternal
+    {
+        public override event EventHandler<EditorSettingsChangedEventArgs> Changed;
+
+        private readonly EditorSettingsManager _editorSettingsManager;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+
+        public DefaultEditorSettingsManagerInternal(ForegroundDispatcher foregroundDispatcher, EditorSettingsManager editorSettingsManager)
+        {
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (editorSettingsManager == null)
+            {
+                throw new ArgumentNullException(nameof(editorSettingsManager));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _editorSettingsManager = editorSettingsManager;
+            _editorSettingsManager.Changed += OnChanged;
+        }
+
+        public override EditorSettings Current => _editorSettingsManager.Current;
+
+        // Internal for testing
+        internal void OnChanged(object sender, EditorSettingsChangedEventArgs e)
```

Comment:
This usage here has a very serious bug. If you're using weakevents like this, you need to keep a reference to the **delegate**.

That looks like:
```
private readonly EventHandler<EditorSettingsChangedEventArgs> _handler;
...

_handler = OnChanged;
_editorSettingsManager.Changed += _handler;
```

The reason is that the thing you have a weakreference to is an `EventHandler<EditorSettingsChangedEventArgs>`. Method-group-to-delegate-conversion will always return a new instance, so you need to manually cache it.

Without manually caching `_handler` in this case the only reference is the weakreference, and it's immediately eligible for GC.

Reply:
Yup you're totally right

Reply:
Thought of an awesome solution. Going to have the `EditorSettingsManagerInternal` track its changed listener. When it has 0 listeners it'll detatch from `EditorSettingsManager`. Basically we're forwarding our lifetime problems to anyone listening to `EditorSettingsManagerInternal`. In this case the document tracker listens to the settings manager internal and it does know when things are tearing down.

This way we can do away with weak references entirely on the `EditorSettingsManager` and we'll still be garbage collected.

Reply:
OK, that sounds like a good solution 👍